### ISSUE TITLE
Optimization Results not Propagated to FOQUS

### ISSUE THREAD

**Issue by boverhof** [CCSI-Toolset/FOQUS#639]
# Description
During execution of a 5 sample optimization using the IPOpt solver @boverhof  discovered that runs were giving identical output, which was unexpected.  On analysis @fsoepyan noticed that the operating variables are not changing from iteration to iteration during the OUU run.

# Details
- Windows 2016 Server / AspenTech v10
- Windows 2008 Server / AspenTech v8.4
- https://github.com/CCSI-Toolset/FOQUS/blob/master/examples/test_files/OUU_Remote_Turbine_BFB/BFB_OUU_08_20_2019.foqus

# Results
- https://github.com/CCSI-Toolset/FOQUS_Sessions/blob/master/examples/test_files/OUU_Remote_Turbine_BFB/BFB_OUU_08_20_2019-Windows2016-Local-5.foqus

**Comment by boverhof** [comment]
# Wrong cost_coe_obj value
In the debug log message from `foqus_lib/framework/graph/node.py` we can see the correct value from ACM.  However in the UI we see FOQUS has the initial value still.

<img width="1380" alt="FOQUS_cost_coe_obj_value" src="https://user-images.githubusercontent.com/738295/69449802-8f1e8000-0d10-11ea-89f4-e802ef5eb7cb.png">

**Comment by eslickj** [comment]
Okay, this is not really a FOQUS bug exactly.  I guess at some point the behavior of ```eval()``` changed, probably with Python 3.  When the nodeScript creates a new f it doesn't replace the local f in the node's runPython function.  The new f is local to what's in eval.  Here is a new node script that fixes the problem:

```python
#run steady state init
self.options["Script"].value = "Run_Steady"
self.runModel()
if self.calcError != -1:
    raise(Exception("Steady state homotopy failed"))
#Run optimization
self.options["Script"].value = "Init_Opt"
self.runModel()
if self.calcError != -1:
    raise(Exception("Optimization failed"))
# Update the x and f dicts from the node output
# f gets copied to the node outputs so need this 
# for now x doesn't get copied back
x, fnew = self.getValues()
for v in f:
	f[v] = fnew[v]
```

**Comment by eslickj** [comment]
I guess to elaborate a bit more, the f dict is supposed to be the node output.  The f dict values get copied back to the node after running the script.  Usually that's find, but when you create a new f dict in eval, f in the eval code points to the new dict and f in the function that contains eval points to the old f.  So the problem was that after running the script the original outputs were being put back in.  This doesn't require FOQUS changes, and should only affect this OUU problem.  I guess it's a documentation thing.

**Comment by boverhof** [comment]
I'm not sure if there is something in FOQUS we could do as a sanity check or this is simply an implementation detail that we can't help exposing to end users?  The concern is it sounds like you need to be a FOQUS expert to use this.

**Comment by eslickj** [comment]
Yeah, you kind of do need to be a FOQUS expert, but we can probably make this a lot easier.  We'll have to think about how we want it to work.

**Comment by fsoepyan** [comment]
- @eslickj: Thank you! I tested the FOQUS BFB-OUU file with the new code, and we no longer have this issue.
- By the way, do we need to modify the documentation?

**Comment by eslickj** [comment]
@fsoepyan we probably should update the docs, but I don't think this stuff is documented anywhere currently.

**Comment by fsoepyan** [comment]
The changes that @eslickj recommended have been merged in #646.
